src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java

```diff
@@ -679,49 +708,52 @@ private String key() {
 
     /** Handles normal build completion (including errors) but also handles the case that the flow did not even start correctly, for example due to an error in {@link FlowExecution#start}. */
     private void finish(@Nonnull Result r, @CheckForNull Throwable t) {
-        setResult(r);
-        duration = Math.max(0, System.currentTimeMillis() - getStartTimeInMillis());
-        LOGGER.log(Level.INFO, "{0} completed: {1}", new Object[] {toString(), getResult()});
-        if (listener == null) {
-            LOGGER.log(Level.WARNING, this + " failed to start", t);
-        } else {
-            RunListener.fireCompleted(WorkflowRun.this, listener);
-            if (t instanceof AbortException) {
-                listener.error(t.getMessage());
-            } else if (t instanceof FlowInterruptedException) {
-                ((FlowInterruptedException) t).handle(this, listener);
-            } else if (t != null) {
-                listener.getLogger().println(Functions.printThrowable(t).trim()); // TODO 2.43+ use Functions.printStackTrace
-            }
-            listener.finished(getResult());
-            listener.closeQuietly();
+        synchronized (getLogCopyGuard()) {
+            setResult(r);
+            completed = Boolean.TRUE;
+            duration = Math.max(0, System.currentTimeMillis() - getStartTimeInMillis());
         }
-        logsToCopy = null;
+
         try {
-            save();
-        } catch (Exception x) {
-            LOGGER.log(Level.WARNING, "failed to save " + this, x);
-        }
-        Timer.get().submit(() -> {
+            LOGGER.log(Level.INFO, "{0} completed: {1}", new Object[]{toString(), getResult()});
+            if (listener == null) {
```

Comment:
what about `getListener()`?

Reply:
Similar to the comment on synchronizing `getListener()` -- give me a bit and I'll either have it changed or better documented in code comments.